File: julia/Scorio.jl/test/rank/test_priors.jl
using Test
using Scorio

@testset "rank/test_priors.jl" begin
    R_small = ordered_binary_small_R()

    @testset "Prior abstract" begin
        @test_throws MethodError Scorio.Rank.Prior()
    end

    @testset "Finite penalties" begin
        theta = [-0.3, -0.1, 0.1, 0.3]

        priors = [
            Scorio.Rank.GaussianPrior(0.0, 1.0),
            Scorio.Rank.LaplacePrior(0.0, 1.0),
            Scorio.Rank.CauchyPrior(0.0, 1.0),
            Scorio.Rank.UniformPrior(),
            Scorio.Rank.CustomPrior(x -> sum(abs.(x))),
            Scorio.Rank.EmpiricalPrior(R_small; var=1.5),
        ]

        for prior in priors
            penalty = Scorio.penalty(prior, theta)
            @test isfinite(Float64(penalty))
        end
    end

    @testset "Empirical prior shape and centering" begin
        prior = Scorio.Rank.EmpiricalPrior(R_small; var=1.0)
        @test length(prior.prior_mean) == size(R_small, 1)
        @test sum(prior.prior_mean) ≈ 0.0 atol = 1e-10
    end

    @testset "Empirical prior theta length validation" begin
        prior = Scorio.Rank.EmpiricalPrior(R_small)
        @test_throws ErrorException Scorio.penalty(prior, [0.0, 1.0])
    end

    @testset "Constructor validation" begin
        @test_throws ErrorException Scorio.Rank.GaussianPrior(0.0, 0.0)
        @test_throws ErrorException Scorio.Rank.LaplacePrior(0.0, 0.0)
        @test_throws ErrorException Scorio.Rank.CauchyPrior(0.0, 0.0)
        @test_throws ErrorException Scorio.Rank.EmpiricalPrior(zeros(Int, 4, 3); var=0.0)
        @test_throws ErrorException Scorio.Rank.EmpiricalPrior(zeros(Int, 2, 3, 4, 5))
        @test_throws ErrorException Scorio.Rank.CustomPrior(42)
    end
end

File: tests/rank/test_priors.py
from __future__ import annotations

import numpy as np
import pytest

from scorio import rank


def test_prior_is_abstract() -> None:
    with pytest.raises(TypeError):
        rank.Prior()


def test_prior_penalty_outputs_are_finite(ordered_binary_small_R: np.ndarray) -> None:
    theta = np.array([-0.3, -0.1, 0.1, 0.3], dtype=float)

    priors = [
        rank.GaussianPrior(mean=0.0, var=1.0),
        rank.LaplacePrior(loc=0.0, scale=1.0),
        rank.CauchyPrior(loc=0.0, scale=1.0),
        rank.UniformPrior(),
        rank.CustomPrior(lambda x: float(np.sum(np.abs(x)))),
        rank.EmpiricalPrior(ordered_binary_small_R, var=1.5),
    ]

    for prior in priors:
        penalty = float(prior.penalty(theta))
        assert np.isfinite(penalty)


def test_empirical_prior_centering_and_shape(
    ordered_binary_small_R: np.ndarray,
) -> None:
    prior = rank.EmpiricalPrior(ordered_binary_small_R, var=1.0)
    assert prior.prior_mean.shape == (ordered_binary_small_R.shape[0],)
    assert float(np.sum(prior.prior_mean)) == pytest.approx(0.0, abs=1e-10)


def test_empirical_prior_rejects_wrong_theta_length(
    ordered_binary_small_R: np.ndarray,
) -> None:
    prior = rank.EmpiricalPrior(ordered_binary_small_R)
    with pytest.raises(ValueError, match="must match number of models"):
        prior.penalty(np.array([0.0, 1.0], dtype=float))


@pytest.mark.parametrize(
    ("ctor", "kwargs", "match"),
    [
        (rank.GaussianPrior, {"var": 0.0}, "Variance must be positive"),
        (rank.LaplacePrior, {"scale": 0.0}, "Scale must be positive"),
        (rank.CauchyPrior, {"scale": 0.0}, "Scale must be positive"),
        (
            rank.EmpiricalPrior,
            {"R0": np.zeros((4, 3), dtype=int), "var": 0.0},
            "Variance must be positive",
        ),
    ],
)
def test_prior_constructor_validation_errors(ctor, kwargs, match: str) -> None:
    with pytest.raises(ValueError, match=match):
        ctor(**kwargs)


def test_empirical_prior_rejects_invalid_R0_shape() -> None:
    with pytest.raises(ValueError, match=r"R0 must be 2D \(L, M\) or 3D \(L, M, D\)"):
        rank.EmpiricalPrior(np.zeros((2, 3, 4, 5), dtype=int))


def test_custom_prior_requires_callable() -> None:
    with pytest.raises(ValueError, match="penalty_fn must be callable"):
        rank.CustomPrior(penalty_fn=42)
